identity-iota/src/tangle/resolver.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+#[cfg(not(feature = "arc"))]
+use std::rc::Rc;
+#[cfg(feature = "arc")]
+use std::sync::Arc;
+
+use dashmap::DashMap;
+
+use identity_core::common::Url;
+use identity_credential::credential::Credential;
+use identity_credential::presentation::Presentation;
+
+use crate::chain::ChainHistory;
+use crate::chain::DocumentHistory;
+use crate::did::IotaDID;
+use crate::diff::DiffMessage;
+use crate::document::ResolvedIotaDocument;
+use crate::error::Error;
+use crate::error::Result;
+use crate::tangle::Client;
+use crate::tangle::ClientBuilder;
+use crate::tangle::NetworkName;
+use crate::tangle::TangleResolve;
+
+#[cfg(not(feature = "arc"))]
+type ClientRc = Rc<Client>;
+
+#[cfg(feature = "arc")]
+type ClientRc = Arc<Client>;
+
+/// A `Resolver` supports resolving DID Documents across different Tangle networks using
+/// multiple [`Clients`][Client].
+///
+/// Also provides convenience functions for resolving DID Documents associated with
+/// verifiable [`Credentials`][Credential] and [`Presentations`][Presentation].
+#[derive(Debug)]
+pub struct Resolver {
+  // TODO: is DashMap really necessary here?
+  //  We can avoid mutation if we disallow adding/removing Clients post-creation.
+  client_map: DashMap<NetworkName, ClientRc>,
+}
+
+impl Resolver {
+  /// Constructs a new [`Resolver`] with a default [`Client`] for
+  /// the [`Mainnet`](crate::tangle::Network::Mainnet).
+  ///
+  /// See also [`Resolver::builder`].
+  pub async fn new() -> Result<Self> {
+    let client_map: DashMap<NetworkName, ClientRc> = DashMap::new();
+    let client: Client = Client::new().await?;
+    client_map.insert(client.network.name(), ClientRc::new(client));
+
+    Ok(Self { client_map })
+  }
+
+  /// Returns a new [`ResolverBuilder`] with no configured [`Clients`](Client).
+  pub fn builder() -> ResolverBuilder {
+    ResolverBuilder::new()
+  }
+
+  /// Inserts a [`Client`].
+  ///
+  /// NOTE: replaces any existing [`Client`] entry with the same [`NetworkName`].
+  // TODO: remove this?
+  pub fn set_client(&self, client: ClientRc) {
+    self.client_map.insert(client.network.name(), client);
+  }
+
+  /// Removes the [`Client`] corresponding to the given [`NetworkName`] and returns it
+  /// if one exists.
+  // TODO: remove this?
+  pub fn remove_client(&self, network_name: &NetworkName) -> Option<ClientRc> {
+    self.client_map.remove(network_name).map(|(_, client)| client)
+  }
+
+  /// Returns the [`Client`] corresponding to the given [`NetworkName`] if one exists.
+  pub fn get_client(&self, network_name: &NetworkName) -> Option<ClientRc> {
```

Comment:
Consider the following scenario:  
Crate x: depends on identity-iota without the `arc` feature 
in this crate there is a function 
```
pub fn get_client() -> Option<Rc<Client>> {
  let resolver: Resolver = .. ;
  resolver.get_client()
}
```
Crate y: depends on identity-iota with the `arc` feature 
Crate z: depends directly on Crate x and has a function 
`fn foo(client: Rc<Client>); `
The following code works in crate z: 
```
let client = crate_x::get_client();
foo(client);
```
but if crate y ever becomes part of crate z's dependency tree (possibly via transitive dependencies) I think crate `z`will no longer compile? 

This is probably the reason why the `im` crate comes in two flavours `im` and `im-rc` (See https://docs.rs/im/latest/im/#thread-safety) instead of changing the pointer types via feature flags. 

We might want to consider using the [`archery crate`](https://docs.rs/archery/latest/archery/) and `Resolver` generic over the pointer type. Then the `new` constructor in the `WasmResolver` can use the `Rc` variant and in the Rust library we either remove `new` and require construction through the builder or use an `Arc` there.

Reply:
> Dependencies are resolved with the union of all features enabled on them. For example, if one package depends on the [im](https://crates.io/crates/im) package with the serde dependency enabled and another package depends on it with the rayon dependency enabled, then im will be built with both features enabled, and the serde and rayon crates will be included in the resolve graph. If no packages depend on im with those features, then those optional dependencies will be ignored, and they will not affect resolution. https://doc.rust-lang.org/cargo/reference/resolver.html#features

I guess you're right that if our crate is a dependency of two other crates, and one enables `arc` and the other doesn't, there will be a conflict.

The `archery` crate looks interesting.

Reply:
Great point, I agree that putting `arc` as a feature isn't ideal since it's not additive and can lead to the situation you described.

I'll try `archery` for now since we would want to use the same pattern to fix the `Account` using `Arc` instead of `Rc` in Wasm.

Reply:
So the problem with `archery` I've found is that it does not support converting from `Arc`/`Rc` to their `SharedPointer`, nor vice-versa (at least as far as I can tell).

This is a problem because we explicitly use `Rc<Client>` in `WasmClient` rather than `SharedPointer<Client, RcK>` (which the `Resolver` uses), so it's not possible to convert the `Resolver` entries into `WasmClient`.

E.g.
```rust 
#[wasm_bindgen(js_name = getClient)]
pub fn get_client(&self, network_name: String) -> Option<WasmClient> {
  let network_name: NetworkName = NetworkName::try_from(network_name).ok()?;
  let client: SharedPointer<Client, RcK> = self.0.get_client(&network_name).cloned()?;
  let rc_client: Rc<Client> = client.into(); // not possible with any safe function archery exposes.
  Some(WasmClient { client: rc_client })
}
```

The reverse also appears not possible.

E.g.
```rust
#[wasm_bindgen]
pub fn client(mut self, client: &WasmClient) -> WasmResolverBuilder {
  let client: Rc<Client> = client.client.clone();
  let shared_client: SharedPointer<Client, RcK> = client.into(); // not possible with any safe function archery exposes.
  self.0 = self.0.client(shared_client);
  self
}
```

Reply:
One solution would be to use `SharedPointer<Client, RcK>` exclusively in the Wasm bindings though, if we want to rely on `archery` completely? 

I don't think this is very friendly to Rust developers though, since I imagine they might want something like a base `Arc<Client>` eventually.

Reply:
```
let shared_client: SharedPointer<Client, RcK> = client.into(); // not possible with any safe function archery exposes.
```

Can't you use https://docs.rs/archery/latest/archery/shared_pointer/struct.SharedPointer.html#method.new ?

Reply:
`SharedPointer::new` takes `T`, not `Rc<T>`.

Reply:
Can we instead of using `archery` just use the sealed trait pattern like so:

```
mod private {
  use super::Client;
  pub(super) trait Sealed {}
  impl Sealed for std::sync::Arc<Client> {}
  impl Sealed for std::rc::Rc<Client> {}
}

```
pub struct Resolver<T: private::Sealed> {
  // TODO: is DashMap really necessary here?
  //  We can avoid mutation if we disallow adding/removing Clients post-creation.
  client_map: DashMap<NetworkName, T>,
}
```
?

Reply:
Yeah there's two approaches I'm trying to hack together now. 

1. Your suggestion of using `Arc<Client>`/`Rc<Client>` as the generic itself. 
2. A trait with an associated type for `Rc` or `Arc`. https://github.com/orium/rpds/issues/7#issuecomment-362635901 , https://joshlf.com/post/2018/10/18/rust-higher-kinded-types-already/#comment-4160863400

Reply:
This is what I have currently. It works and doesn't require any odd hacks. 

We can restrict it using a sealed trait to prevent unexpected entries for `T` from dependents, but I don't think it's too bad as-is?

```rust
pub struct Resolver<T = Arc<Client>>
where
  T: Clone + AsRef<Client> + From<Client>,
{
  client_map: HashMap<NetworkName, T>,
}
```

```rust
// Developers using it in Rust will default to Resolver<Arc<Client>.
let resolver: Resolver = Resolver::new().await?; 

// Changing to Rc is a bit verbose but not bad.
let resolver: Resolver<Rc<Client>> = Resolver::<Rc<Client>>::new().await?;
```

Using a generic trait with an associated type of `Rc` or `Arc` for a generic `T` (rather than just `Client`) is another option but unnecessary in our case since we only require it for `Client`, in my opinion.

Reply:
I think that's perfect :+1:

Reply:
Actually, I think you're right that a sealed trait might be better so we can change it without affecting semantic versioning.

Reply:
Thanks for all the feedback. I'll leave it as-is.